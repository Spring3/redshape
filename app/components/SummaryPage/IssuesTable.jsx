import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _get from 'lodash/get';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import styled, { withTheme, css } from 'styled-components';
import SortAscendingIcon from 'mdi-react/SortAscendingIcon';
import SortDescendingIcon from 'mdi-react/SortDescendingIcon';

import Table from '../Table';

const ColorfulSpan = styled.span`
  ${
    props => props.color
    ? css `
      padding-bottom: 2px;
      background: linear-gradient(to bottom,transparent 0,transparent 90%,${props.color} 90%,${props.color} 100%);
    `
    : null
  }
`;

const colorMap = {
  'closed': 'red',
  'high': 'red',
  'open': 'green',
  'low': 'green',
  'pending': 'yellow',
  'normal': 'yellow'
};

class IssuesTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortBy: undefined,
      sortDirection: undefined
    };
  }

  sortTable = (by) => { 
    return () => {
      const { onSort } = this.props;
      const { sortBy, sortDirection } = this.state;
      if (sortBy !== by) {
        this.setState({
          sortBy: by,
          sortDirection: 'asc'
        }, () => onSort(by, 'asc'));
      } else {
        const newSortDirection = sortDirection === 'asc' ? 'desc' : 'asc';
        this.setState({ sortDirection: newSortDirection }, () => onSort(by, newSortDirection));
      }
    };
  }

  showIssueDetails(id) {
    this.props.history.push(`/app/issue/${id}/`);
  }

  paint = (item, mapping) => {
    const { theme, useColors } = this.props;
    const textValue = _get(item, mapping);
    
    const color = (useColors && typeof textValue === 'string'
      ? colorMap[textValue.toLowerCase()]
      : undefined);
    return (
      <ColorfulSpan color={theme[color]}>{textValue}</ColorfulSpan>
    );
  }

  render() {
    const { issueHeaders, userTasks } = this.props;
    const { sortBy, sortDirection } = this.state;
    return (
      <Table>
        <tr>
          {issueHeaders.map(header => (
            <th
              key={header.value}
              className={header.value === 'due_date' ? 'due-date' : ''}
              onClick={this.sortTable(header.value)}
            >
              {header.label}&nbsp;
                  {sortBy === header.value && sortDirection === 'asc' && (
                <SortAscendingIcon size={14} />
              )}
              {sortBy === header.value && sortDirection === 'desc' && (
                <SortDescendingIcon size={14} />
              )}
            </th>
          ))}
        </tr>
        {userTasks.map(task => (
          <tr key={task.id} onClick={this.showIssueDetails.bind(this, task.id)}>
            {
              issueHeaders.map(header => (
                <td key={header.value}>
                  {this.paint(task, header.value)}
                </td>
              ))
            }
          </tr>
        ))}
      </Table>
    );
  }
}

IssuesTable.propTypes = {
  userTasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  useColors: PropTypes.bool.isRequired,
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  onSort: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  useColors: state.settings.useColors,
  userTasks: state.issues.all.data,
  issueHeaders: state.settings.issueHeaders,
});

export default withRouter(withTheme(connect(mapStateToProps)(IssuesTable)));
