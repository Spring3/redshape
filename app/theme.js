export default {
  bg: '#FFFFFF',
  bgLight: '#F9F9F9',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#909090',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  transitionTime: '.2s'
};
