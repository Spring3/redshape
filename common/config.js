const storage = require('./storage');
const env = require('./env');

let config = { ...env };

if (storage.has('settings')) {
  config = { ...config, ...storage.get('settings') };
}

module.exports = config;
