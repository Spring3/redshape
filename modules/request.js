const axios = require('axios');
const storage = require('./storage');

let instance;

const TWENTY_SECONDS = 20000;

const pendingRequests = {};

const makeCancellable = (headers, requestId) => {
  if (!requestId) {
    return headers;
  }

  if (pendingRequests[requestId]) {
    pendingRequests[requestId].cancel();
  }
  const source = axios.CancelToken.source();
  const newHeaders = {
    ...headers,
    cancelToken: source.token
  };
  pendingRequests[requestId] = source;
  return newHeaders;
};

const initialize = (redmineEndpoint, token) => {
  if (!redmineEndpoint || !token) {
    return;
  }
  const axiosConfig = {
    baseURL: redmineEndpoint,
    timeout: TWENTY_SECONDS,
    headers: {
      'X-Redmine-API-Key': token
    }
  };
  instance = axios.create(axiosConfig);
};

const getInstance = () => instance;

const reset = () => { instance = undefined; };

if (storage.has('user')) {
  const { api_key, redmineEndpoint } = storage.get('user');
  if (api_key && redmineEndpoint) {
    initialize(redmineEndpoint, api_key);
  }
}

const handleReject = (error) => {
  // if this request was not cancelled
  if (!axios.isCancel(error)) {
    let errorMessage = 'Error';
    if (error.status) {
      errorMessage = `${errorMessage} ${error.status}`;
    }
    errorMessage = `${errorMessage} (${error.statusText || error.message})`;

    return Promise.reject(new Error(errorMessage));
  }
  return undefined;
};

const authorizedRequest = (config) => {
  if (config.id) {
    Object.assign(
      config,
      { headers: makeCancellable(config.headers || {}, config.id) }
    );
  }
  if (!instance) {
    return Promise.reject(new Error('401 - Unauthorized'));
  }

  Object.assign(
    config,
    {
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
        'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
        'X-Redmine-API-Key': instance.defaults.headers['X-Redmine-API-Key'],
        ...config.headers
      }
    }
  );

  return getInstance().request(config)
    .then((res) => {
      delete pendingRequests[config.id];
      return ({ data: res.data });
    })
    .catch((error) => {
      delete pendingRequests[config.id];
      return handleReject(error);
    });
};

const request = (config) => {
  Object.assign(config, {
    timeout: TWENTY_SECONDS,
    headers: {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, PATCH, OPTIONS',
      'Access-Control-Allow-Headers': 'X-Requested-With, content-type, Authorization',
      ...config.headers
    }
  });
  return axios.request(config).catch(handleReject);
};

module.exports = {
  default: axios,
  makeCancellable,
  initialize,
  getInstance,
  pendingRequests,
  reset,
  authorizedRequest,
  request
};
