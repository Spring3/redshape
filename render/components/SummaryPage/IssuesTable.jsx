import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import _get from 'lodash/get';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import styled, { withTheme, css } from 'styled-components';
import SortAscendingIcon from 'mdi-react/SortAscendingIcon';
import SortDescendingIcon from 'mdi-react/SortDescendingIcon';

import InfiniteScroll from '../InfiniteScroll';
import ProcessIndicator, { OverlayProcessIndicator } from '../ProcessIndicator';
import Date from '../Date';
import Progressbar from "../Progressbar";

import { Tag, Status, IssueId, Priority } from "../Issue";

const Table = styled.table`
  position: relative;
  min-height: 200px;
  width: 100%;
  border: 2px solid ${props => props.theme.bgDark};
  border-spacing: 0px;

  tbody {
    text-align: center;
    font-size: 14px;
    overflow-y: scroll;
    max-height: 60vh;

    tr {
      td {
        border-bottom: 2px solid ${props => props.theme.bgDark};
        padding: 15px 5px;
        transition: background ease ${props => props.theme.transitionTime};
      }

      &:hover {
        cursor: pointer;
        
        td {
          background: ${props => props.theme.bgDark};
        }
      }

      td.due_date {
        white-space: nowrap;
      }
      
      td.subject {
        text-align: left;
      }
    }
  }

  thead {
    z-index: 1;
    tr {
      th {
        padding: 15px 5px;
        background: ${props => props.theme.bgDark};
        border-bottom: 2px solid ${props => props.theme.bgDarker};
        transition: color ease ${props => props.theme.transitionTime};
        &:hover {
          cursor: pointer;
          color: ${props => props.theme.main};
        }
  
        svg {
          vertical-align: middle;
        }
        
        &.subject {
          text-align: left;
        }
      }
    }
  }
`;

const ProcessIndicatorContainer = styled.tr`
  position: absolute;
  width: 100%;
  text-align: center;

  div.container {
    position: absolute;
    top: 15px;
    left: 45%;
  }
`;

const ColorfulSpan = styled.span`
  ${
    props => props.color
    ? css `
      padding-bottom: 2px;
      background: linear-gradient(to bottom,transparent 0,transparent 90%,${props.color} 90%,${props.color} 100%);
    `
    : null
  }
`;

const colorMap = {
  'closed': 'red',
  'high': 'yellow',
  'urgent': 'red',
  'immediate': 'red',
  'critical': 'red',
  'open': 'green',
  'low': 'green',
  'pending': 'yellow',
  'normal': 'yellow'
};

class IssuesTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sortBy: undefined,
      sortDirection: undefined
    };
  }

  sortTable = (by) => () => {
    const { onSort } = this.props;
    const { sortBy, sortDirection } = this.state;
    if (sortBy !== by) {
      this.setState({
        sortBy: by,
        sortDirection: 'asc'
      }, () => onSort(by, 'asc'));
    } else {
      const newSortDirection = sortDirection === 'asc' ? 'desc' : 'asc';
      this.setState({ sortDirection: newSortDirection }, () => onSort(by, newSortDirection));
    }
  }

  showIssueDetails(id) {
    this.props.history.push(`/app/issue/${id}/`);
  }

  paint = (format, value, textValue, opts) => {
    const { theme, uiStyle } = this.props;
    const isEnhanced = opts.isEnhanced;

    if (isEnhanced){
      if (format === 'id' && isEnhanced) {
        return (<IssueId value={value} tracker={opts.tracker}/>);
      }else if (format === 'status' && isEnhanced) {
        return (<Status simple={true} value={value}/>);
      }else if (format === 'progress' && isEnhanced) {
        return (<Progressbar percent={value} width={80} mode="progress-gradient" height={7}/>);
      }else if (format === 'priority' && isEnhanced) {
        return (<Priority centered value={value}/>);
      }
    }

    if (format === 'date'){
      return (<Date date={value}/>);
    }else if (format === 'tags' && value){
      return (<Fragment>{ value ? value.map(el => <Tag key={el} mode={isEnhanced ? "simple" : "plain"} value={el}></Tag>) : undefined}</Fragment>);
    }else{
      const color = (uiStyle === 'colors' && typeof value === 'string'
        ? colorMap[value.toLowerCase()]
        : undefined);

      return (
        <ColorfulSpan color={theme[color]} className={opts && opts.className}>{textValue}</ColorfulSpan>
      );
    }
  }

  loadIssuePage = () => {
    const { issues, fetchIssuePage } = this.props;
    const { page } = issues;
    fetchIssuePage(page + 1);
  }

  render() {
    const { issueHeaders, issues, limit, uiStyle } = this.props;
    const { sortBy, sortDirection } = this.state;
    let userTasks = issues.data;
    if (limit){
      userTasks = userTasks.slice(0,  limit)
    }
    const isEnhanced = uiStyle === 'enhanced';
    return (
      <Fragment>
        { (!userTasks.length && issues.isFetching) && (<OverlayProcessIndicator />) }
        <Table>
          <thead>
          <tr>
            {issueHeaders.map(header => (
              <th
                key={header.value}
                onClick={this.sortTable(header.value)}
                className={isEnhanced ? header.value : ""}
              >
                {header.label}&nbsp;
                {sortBy === header.value && sortDirection === 'asc' && (
                  <SortAscendingIcon size={14} />
                )}
                {sortBy === header.value && sortDirection === 'desc' && (
                  <SortDescendingIcon size={14} />
                )}
              </th>
            ))}
          </tr>
          </thead>
          <tbody>
          <InfiniteScroll
            load={this.loadIssuePage}
            isEnd={userTasks.length === issues.totalCount || limit === userTasks.length}
            hasMore={!issues.isFetching && !issues.error && userTasks.length < issues.totalCount && !limit}
            container={window}
            loadIndicator={<ProcessIndicatorContainer><td><ProcessIndicator className="container" /></td></ProcessIndicatorContainer>}
            immediate={true}
          >
            {userTasks.map((task)  => (
              <tr key={task.id} onClick={this.showIssueDetails.bind(this, task.id)}>
                {
                  issueHeaders.map(header => {
                    let opts = {};
                    if (isEnhanced) {
                      opts.isEnhanced = isEnhanced;
                      opts.tracker = _get(task, 'tracker.id');
                    }
                    const format = header.format;
                    let value = _get(task, header.value);
                    let textValue = value;
                    if (format === 'hours'){
                      if (value != null){
                        textValue = Number(value.toFixed(2));
                      }
                    }else if (format === 'progress'){
                      textValue = `${value}%`
                    }else if (format === 'count' && value && value.length){
                      textValue = `${value.length}`
                    }
                    return (
                      <td
                        key={header.value}
                        className={isEnhanced ? header.value : (header.value === 'due_date' ? header.value : "")}
                      >
                        { this.paint(format, value, textValue, opts) }
                      </td>
                    )})
                }
              </tr>
            ))}
          </InfiniteScroll>
          </tbody>
        </Table>
      </Fragment>
    );
  }
}

IssuesTable.propTypes = {
  issues: PropTypes.shape({
    userTasks: PropTypes.arrayOf(PropTypes.object).isRequired,
  }).isRequired,
  uiStyle: PropTypes.string.isRequired,
  issueHeaders: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    isFixed: PropTypes.bool
  }).isRequired).isRequired,
  onSort: PropTypes.func.isRequired,
  fetchIssuePage: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  uiStyle: state.settings.uiStyle,
  issues: state.issues.all,
  issueHeaders: state.settings.issueHeaders,
});

export default withRouter(withTheme(connect(mapStateToProps)(IssuesTable)));
