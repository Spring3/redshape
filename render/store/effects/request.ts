import { ipcRenderer, remote } from 'electron';
import { Response } from '../../../types';

const crypto = remote.require('crypto');

type Config = {
  endpoint: string;
  token?: string;
}

type QueryConfig = {
  payload: Record<string, any>;
  config?: Config;
}

const query = (config: QueryConfig): Promise<Response> => {
  const id = crypto.randomBytes(10).toString('hex');

  return new Promise((resolve) => {
    ipcRenderer.send('request', JSON.stringify({ ...config, id }));

    ipcRenderer.once(`response:${id}`, (event, response) => {
      console.log('response:', response);
      resolve(response);
    });
  });
};

export {
  query,
};

export type {
  QueryConfig
};
