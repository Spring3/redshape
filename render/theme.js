export default {
  bg: '#FFFFFF',
  bgDark: '#F9F9F9',
  bgDarker: '#E9E9E9',
  bgDisabled: 'hsl(0,0%,95%)',
  mainLight: '#FF7F87',
  main: '#FF7079',
  mainDark: '#EF6069',
  minorText: '#A0A0A0',
  normalText: '#3F3844',
  hoverText: '#FFFFFF',
  red: '#FF634D',
  yellow: '#FFDA77',
  green: '#6CCA51',
  shadow: '#D0D0D0',
  transitionTime: '.2s'
};
