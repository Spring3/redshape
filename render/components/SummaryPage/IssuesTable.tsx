import React, {
  useState, useMemo, useCallback
} from 'react';
import { css } from '@emotion/react';
import { useHistory } from 'react-router-dom';
import { useTheme } from 'styled-components';

import { get } from 'lodash';
import { theme as Theme } from '../../theme';
import { ProcessIndicator, OverlayProcessIndicator } from '../ProcessIndicator';
import Date from '../Date';
import { useOvermindActions, useOvermindState } from '../../store';
import { IssueFilter } from '../../actions/helper';
import { usePaginatedFetch } from '../../hooks/usePaginatedFetch';
import { issueHeaders } from '../constants';
import { IssuesTableHead } from './IssuesTableHead';
import { IssueHeader, SortingDirection } from '../../../types';
import { IssuesTableEmptyState } from './IssuesTableEmptyState';
import { IssuesTableEmptySearchState } from './IssuesTableEmptySearchState';

const styles = {
  processIndicatorText: css`
    white-space: nowrap;
    padding-left: 20px;
    vertical-align: middle;
  `,
  tableBody: css`
    text-align: center;
    font-size: 14px;
    overflow-y: scroll;
    max-height: 60vh;
  `,
  processIndicatorWrapper: css`
    position: absolute;
    width: 100%;
    display: flex;
    margin-top: 1.5rem;
  `,
  processIndicatorContainer: css`
    margin: 0 auto;
  `
};

const IssuesTable = ({ search }: { search?: string }) => {
  const [sortBy, setSortBy] = useState<IssueHeader['value']>();
  const [sortDirection, setSortDirection] = useState<SortingDirection>('asc');

  const history = useHistory();
  const state = useOvermindState();
  const actions = useOvermindActions();

  const theme = useTheme() as typeof Theme;

  const filters = useMemo(
    () => new IssueFilter()
      .assignee(state.users.currentUser?.id as string)
      .status({ open: true, closed: state.settings.showClosedIssues })
      .title(search)
      .sort(sortBy, sortDirection)
      .build(),
    [state.users.currentUser, state.settings.showClosedIssues, search, sortBy, sortDirection]
  );

  const {
    isFetching, items: issues, error, fetchTriggerElementRef
  } = usePaginatedFetch({
    request: actions.issues.getMany,
    filters,
  });

  const sortTable = useCallback((by: IssueHeader['value']) => {
    if (sortBy !== by) {
      setSortBy(by);
      setSortDirection('asc');
    } else {
      setSortDirection(sortDir => (sortDir === 'asc' ? 'desc' : 'asc'));
    }
  }, [sortBy]);

  const showIssueDetails = (id: number) => {
    history.push(`/app/issue/${id}/`);
  };

  if (!issues.length && isFetching) {
    return (
      <OverlayProcessIndicator>
        <span css={styles.processIndicatorText}>Please wait...</span>
      </OverlayProcessIndicator>
    );
  }

  if (!issues.length && !search) {
    return (
      <>
        <table css={css`
          position: relative;
          width: 100%;
          margin-top: 0.5rem;
          border: 2px solid ${theme.bgDark};
          border-spacing: 0px;
        `}
        >
          <IssuesTableHead />
        </table>
        <IssuesTableEmptyState />
      </>
    );
  }

  if (!issues.length && search) {
    return (
      <>
        <table css={css`
          position: relative;
          width: 100%;
          margin-top: 0.5rem;
          border: 2px solid ${theme.bgDark};
          border-spacing: 0px;
        `}
        >
          <IssuesTableHead />
        </table>
        <IssuesTableEmptySearchState />
      </>
    );
  }

  return (
    <table css={css`
    position: relative;
    width: 100%;
    margin-top: 0.5rem;
    border: 2px solid ${theme.bgDark};
    border-spacing: 0px;
  `}>
      <IssuesTableHead sortBy={sortBy} sortDirection={sortDirection} onSort={sortTable} />
      <tbody css={styles.tableBody}>
        {issues.map((task, index) => (
          <tr
            css={css`
              &:hover {
                cursor: pointer;
          
                td {
                  background: ${theme.bgDark};
                }
              }
            `}
            ref={index === issues.length - 1 ? fetchTriggerElementRef : null}
            key={task.id}
            onClick={() => showIssueDetails(task.id)}
          >
            {issueHeaders.map(header => {
              const date = header.value === 'due_date' && get(task, header.value);
              let forcedValue;
              const estimated_hours = header.value === 'estimated_hours' && get(task, header.value);
              if (estimated_hours) {
                forcedValue = Number(estimated_hours.toFixed(2));
              }
              return (
                <td
                  css={
                  css`
                    cursor: pointer;
                    border-bottom: 2px solid ${theme.bgDark};
                    padding: 15px 5px;
                    transition: background ease ${theme.transitionTime};
                    ${header.value === 'due_date' ? 'white-space: nowrap' : ''};
                    ${header.value === 'subject' ? 'text-align: start' : ''};
                  `
                }
                  key={header.value}
                >
                  {date ? (
                    <Date date={date} />
                  ) : (
                    <span>{forcedValue || get(task, header.value)}</span>
                  )}
                </td>
              );
            })}
          </tr>
        ))}
        {isFetching ? (
          <tr css={styles.processIndicatorWrapper}>
            <td css={styles.processIndicatorContainer}>
              <ProcessIndicator className="container">
                <span css={styles.processIndicatorText}>Fetching more...</span>
              </ProcessIndicator>
            </td>
          </tr>
        ) : null}
      </tbody>
    </table>
  );
};

export { IssuesTable };
