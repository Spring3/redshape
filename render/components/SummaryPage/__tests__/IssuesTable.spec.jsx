import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { HashRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';

import IssuesTable from '../IssuesTable';
import { getInstance, reset, initialize } from '../../../../common/request';
import theme from '../../../theme';

const redmineEndpoint = 'redmint.test.test';
const token = 'multipass';
const mockStore = configureStore([thunk]);
const state = {
  settings: {
    useColors: true,
    issueHeaders: [
      { label: 'Id', isFixed: true, value: 'id' },
      { label: 'Project', value: 'project.name' },
      { label: 'Tracker', value: 'tracker.name' },
      { label: 'Status', value: 'status.name' },
      { label: 'Subject', isFixed: true, value: 'subject' },
      { label: 'Priority', value: 'priority.name' },
      { label: 'Estimation', value: 'estimated_hours' },
      { label: 'Due Date', value: 'due_date' }
    ]
  },
  issues: {
    all: {
      data: [
        {
          id: 1,
          project: { name: 'Test Project 1' },
          tracker: { name: 'Test Tracker 1' },
          status: { name: 'Open' },
          subject: 'Task #1',
          priority: { name: 'High' },
          estimated_hours: '10',
          due_date: '2011-01-01'
        },
        {
          id: 2,
          project: { name: 'Test Project 2' },
          tracker: { name: 'Test Tracker 2' },
          status: { name: 'Open' },
          subject: 'Task #2',
          priority: { name: 'Normal' },
          estimated_hours: '5',
          due_date: '2011-01-01'
        },
        {
          id: 3,
          project: { name: 'Test Project 3' },
          tracker: { name: 'Test Tracker 3' },
          status: { name: 'Open' },
          subject: 'Task #3',
          priority: { name: 'Low' },
          estimated_hours: '2',
          due_date: '2011-01-01'
        }
      ]
    }
  }
};

let axiosMock;

describe('SummaryPage => IssuesTable component', () => {
  beforeAll(() => {
    initialize(redmineEndpoint, token);
    axiosMock = new MockAdapter(getInstance());
    axiosMock.onAny('*').reply(() => Promise.resolve([200]));
  });

  afterEach(() => {
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
    reset();
  });

  it('should match the snapshot', () => {
    const store = mockStore(state);
    const onSort = jest.fn();
    const tree = renderer.create(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssuesTable onSort={onSort} />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('allow the table to be sorted', (done) => {
    const store = mockStore(state);
    const onSort = jest.fn();
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter>
          <ThemeProvider theme={theme}>
            <IssuesTable onSort={onSort} />
          </ThemeProvider>
        </HashRouter>
      </Provider>
    );
    // 1st row, 1st column
    wrapper.find('tr').at(0).find('th').at(0).simulate('click');
    setImmediate(() => {
      expect(onSort).toHaveBeenCalledWith('id', 'asc');
      done();
    });
  });
});
