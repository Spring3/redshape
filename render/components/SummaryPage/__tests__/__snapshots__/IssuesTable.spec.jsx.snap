// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryPage => IssuesTable component should match the snapshot 1`] = `
.c0 {
  position: relative;
  min-height: 200px;
  width: 100%;
  border: 2px solid #F9F9F9;
  border-spacing: 0px;
}

.c0 tbody {
  text-align: center;
  font-size: 14px;
  overflow-y: scroll;
  max-height: 60vh;
}

.c0 tbody tr td {
  border-bottom: 2px solid #F9F9F9;
  padding: 15px 5px;
  -webkit-transition: background ease .2s;
  transition: background ease .2s;
}

.c0 tbody tr:hover {
  cursor: pointer;
}

.c0 tbody tr:hover td {
  background: #F9F9F9;
}

.c0 tbody tr td.due_date {
  white-space: nowrap;
}

.c0 thead {
  z-index: 1;
}

.c0 thead tr th {
  padding: 15px 5px;
  background: #F9F9F9;
  border-bottom: 2px solid #E9E9E9;
  -webkit-transition: color ease .2s;
  transition: color ease .2s;
}

.c0 thead tr th:hover {
  cursor: pointer;
  color: #FF7079;
}

.c0 thead tr th svg {
  vertical-align: middle;
}

.c1 {
  padding-bottom: 2px;
  background: linear-gradient(to bottom,transparent 0,transparent 90%,#6CCA51 90%,#6CCA51 100%);
}

.c2 {
  padding-bottom: 2px;
  background: linear-gradient(to bottom,transparent 0,transparent 90%,#FFDA77 90%,#FFDA77 100%);
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Id
         
      </th>
      <th
        onClick={[Function]}
      >
        Project
         
      </th>
      <th
        onClick={[Function]}
      >
        Tracker
         
      </th>
      <th
        onClick={[Function]}
      >
        Status
         
      </th>
      <th
        onClick={[Function]}
      >
        Subject
         
      </th>
      <th
        onClick={[Function]}
      >
        Priority
         
      </th>
      <th
        onClick={[Function]}
      >
        Estimation
         
      </th>
      <th
        onClick={[Function]}
      >
        Total Estimation
         
      </th>
      <th
        onClick={[Function]}
      >
        Spent
         
      </th>
      <th
        onClick={[Function]}
      >
        Total Spent
         
      </th>
      <th
        onClick={[Function]}
      >
        Progress
         
      </th>
      <th
        onClick={[Function]}
      >
        Due Date
         
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span
          className=""
        >
          1
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Project 1
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Tracker 1
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c1"
          color="#6CCA51"
        >
          Open
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Task #1
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c2"
          color="#FFDA77"
        >
          High
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          10
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className="due_date"
      >
        <span
          className=""
        >
          2011-01-01
        </span>
      </td>
    </tr>
    <tr
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span
          className=""
        >
          2
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Project 2
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Tracker 2
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c1"
          color="#6CCA51"
        >
          Open
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Task #2
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c2"
          color="#FFDA77"
        >
          Normal
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          5
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className="due_date"
      >
        <span
          className=""
        >
          2011-01-01
        </span>
      </td>
    </tr>
    <tr
      onClick={[Function]}
    >
      <td
        className=""
      >
        <span
          className=""
        >
          3
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Project 3
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Test Tracker 3
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c1"
          color="#6CCA51"
        >
          Open
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          Task #3
        </span>
      </td>
      <td
        className=""
      >
        <span
          className="c1"
          color="#6CCA51"
        >
          Low
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        >
          2
        </span>
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className=""
      >
        <span
          className=""
        />
      </td>
      <td
        className="due_date"
      >
        <span
          className=""
        >
          2011-01-01
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
